File: lib/portal-search/filter-set.ts
import {
  AMENIDADES,
  DEFAULT_FILTER_SET,
  ESTAGIOS,
  type FilterSet,
  TIPOS_IMOVEL,
  TRANSACOES,
} from "./types"

export type { FilterSet } from "./types"
export { DEFAULT_FILTER_SET } from "./types"

function isStringArray(value: unknown): value is string[] {
  return Array.isArray(value) && value.every((v) => typeof v === "string")
}

function isNumberArray(value: unknown): value is number[] {
  return Array.isArray(value) && value.every((v) => typeof v === "number" && Number.isInteger(v))
}

function nullableNumber(value: unknown): number | null {
  if (value === null || value === undefined || value === "") return null
  const n = Number(value)
  return Number.isFinite(n) ? n : null
}

export function parseFilterSet(input: unknown): FilterSet {
  const raw = (input && typeof input === "object" ? input : {}) as Record<string, unknown>

  const transacao = TRANSACOES.includes(raw.transacao as (typeof TRANSACOES)[number])
    ? (raw.transacao as FilterSet["transacao"])
    : DEFAULT_FILTER_SET.transacao

  const uf =
    typeof raw.uf === "string" && raw.uf.length === 2
      ? raw.uf.toLowerCase()
      : DEFAULT_FILTER_SET.uf

  const cidade =
    typeof raw.cidade === "string" && raw.cidade.trim()
      ? raw.cidade.trim().toLowerCase()
      : DEFAULT_FILTER_SET.cidade

  const bairros = isStringArray(raw.bairros)
    ? raw.bairros.map((b) => b.trim().toLowerCase()).filter(Boolean)
    : []

  const tiposImovel = isStringArray(raw.tiposImovel)
    ? raw.tiposImovel.filter((t): t is FilterSet["tiposImovel"][number] =>
        TIPOS_IMOVEL.includes(t as (typeof TIPOS_IMOVEL)[number])
      )
    : DEFAULT_FILTER_SET.tiposImovel

  const quartos = isNumberArray(raw.quartos)
    ? raw.quartos.filter((n) => n >= 0 && n <= 5)
    : []

  const banheiros = isNumberArray(raw.banheiros)
    ? raw.banheiros.filter((n) => n >= 0 && n <= 5)
    : []

  const vagas = isNumberArray(raw.vagas) ? raw.vagas.filter((n) => n >= 0 && n <= 5) : []

  const suites = isNumberArray(raw.suites) ? raw.suites.filter((n) => n >= 0 && n <= 5) : []

  const amenidades = isStringArray(raw.amenidades)
    ? raw.amenidades.filter((a): a is FilterSet["amenidades"][number] =>
        AMENIDADES.includes(a as (typeof AMENIDADES)[number])
      )
    : []

  const estagio = isStringArray(raw.estagio)
    ? raw.estagio.filter((e): e is FilterSet["estagio"][number] =>
        ESTAGIOS.includes(e as (typeof ESTAGIOS)[number])
      )
    : []

  return {
    transacao,
    uf,
    cidade,
    bairros,
    tiposImovel: tiposImovel.length > 0 ? tiposImovel : DEFAULT_FILTER_SET.tiposImovel,
    quartos,
    banheiros,
    vagas,
    suites,
    precoMin: nullableNumber(raw.precoMin),
    precoMax: nullableNumber(raw.precoMax),
    areaMin: nullableNumber(raw.areaMin),
    areaMax: nullableNumber(raw.areaMax),
    condominioMax: nullableNumber(raw.condominioMax),
    amenidades,
    estagio,
  }
}

export { DEFAULT_FILTER_SET }

File: backend/lib/minha_casa_ai/portal_searches/filter_set.ex
defmodule MinhaCasaAi.PortalSearches.FilterSet do
  @moduledoc false

  @transacoes ~w(venda aluguel)
  @tipos_imovel ~w(
    apartamento casa sobrado cobertura kitnet studio loft flat
    casa_condominio terreno sala_comercial galpao chacara sitio fazenda
  )
  @amenidades ~w(
    piscina churrasqueira academia sacada varanda_gourmet mobiliado portaria_24h
    elevador salao_de_festas playground quadra sauna seguranca_24h aceita_pets
    ar_condicionado armarios_cozinha armarios_quarto proximo_metro
  )
  @estagios ~w(pronto em_construcao na_planta lancamento)

  def transacoes, do: @transacoes
  def tipos_imovel, do: @tipos_imovel
  def amenidades, do: @amenidades
  def estagios, do: @estagios

  def default do
    %{
      "transacao" => "venda",
      "uf" => "sp",
      "cidade" => "sao-paulo",
      "bairros" => [],
      "tiposImovel" => ["apartamento"],
      "quartos" => [],
      "banheiros" => [],
      "vagas" => [],
      "suites" => [],
      "precoMin" => nil,
      "precoMax" => nil,
      "areaMin" => nil,
      "areaMax" => nil,
      "condominioMax" => nil,
      "amenidades" => [],
      "estagio" => []
    }
  end

  def parse(raw) when is_map(raw) do
    parsed =
      default()
      |> Map.merge(stringify_keys(raw))
      |> normalize()

    case validate(parsed) do
      :ok -> {:ok, parsed}
      {:error, reason} -> {:error, reason}
    end
  end

  def parse(_), do: {:error, :invalid_filter_set}

  defp normalize(map) do
    map
    |> put_string("transacao", &(&1 in @transacoes), "venda")
    |> put_string("uf", fn v -> is_binary(v) and String.length(v) == 2 end, "sp")
    |> update_in(["uf"], &String.downcase/1)
    |> put_string("cidade", &is_binary/1, "sao-paulo")
    |> update_in(["cidade"], &slugify/1)
    |> put_string_list("bairros")
    |> put_string_list("tiposImovel", @tipos_imovel)
    |> put_int_list("quartos")
    |> put_int_list("banheiros")
    |> put_int_list("vagas")
    |> put_int_list("suites")
    |> put_string_list("amenidades", @amenidades)
    |> put_string_list("estagio", @estagios)
    |> put_nullable_number("precoMin")
    |> put_nullable_number("precoMax")
    |> put_nullable_number("areaMin")
    |> put_nullable_number("areaMax")
    |> put_nullable_number("condominioMax")
    |> ensure_tipos()
  end

  defp validate(%{"transacao" => t}) when t in @transacoes, do: :ok
  defp validate(_), do: {:error, :invalid_transacao}

  defp ensure_tipos(%{"tiposImovel" => []} = map),
    do: Map.put(map, "tiposImovel", ["apartamento"])

  defp ensure_tipos(map), do: map

  defp put_string(map, key, valid?, default) do
    case Map.get(map, key) do
      v when is_binary(v) ->
        if valid?.(v), do: map, else: Map.put(map, key, default)

      _ ->
        Map.put(map, key, default)
    end
  end

  defp put_string_list(map, key, allowed \\ nil) do
    values =
      map
      |> Map.get(key, [])
      |> List.wrap()
      |> Enum.filter(&is_binary/1)
      |> Enum.map(&String.trim/1)
      |> Enum.reject(&(&1 == ""))
      |> Enum.map(&slugify/1)

    values =
      if allowed do
        Enum.filter(values, &(&1 in allowed))
      else
        values
      end

    Map.put(map, key, values)
  end

  defp put_int_list(map, key) do
    values =
      map
      |> Map.get(key, [])
      |> List.wrap()
      |> Enum.map(&to_int/1)
      |> Enum.reject(&is_nil/1)
      |> Enum.filter(&(&1 >= 0 and &1 <= 5))

    Map.put(map, key, values)
  end

  defp put_nullable_number(map, key) do
    Map.put(map, key, to_float(Map.get(map, key)))
  end

  defp stringify_keys(map) do
    Map.new(map, fn
      {k, v} when is_atom(k) -> {Atom.to_string(k), v}
      {k, v} -> {to_string(k), v}
    end)
  end

  defp slugify(value) when is_binary(value) do
    value
    |> String.downcase()
    |> String.trim()
    |> String.normalize(:nfd)
    |> String.replace(~r/[^a-z0-9\s-]/u, "")
    |> String.replace(~r/\s+/, "-")
    |> String.replace(~r/-+/, "-")
  end

  defp to_int(n) when is_integer(n), do: n
  defp to_int(n) when is_float(n), do: trunc(n)

  defp to_int(n) when is_binary(n) do
    case Integer.parse(n) do
      {i, _} -> i
      :error -> nil
    end
  end

  defp to_int(_), do: nil

  defp to_float(nil), do: nil
  defp to_float(n) when is_number(n), do: n * 1.0

  defp to_float(n) when is_binary(n) do
    case Float.parse(n) do
      {f, _} -> f
      :error -> nil
    end
  end

  defp to_float(_), do: nil
end
